pkg/controller/colocationprofile/colocationprofile_controller.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/spf13/pflag"
+	"golang.org/x/time/rate"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configv1alpha1 "github.com/koordinator-sh/koordinator/apis/config/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/features"
+	utilfeature "github.com/koordinator-sh/koordinator/pkg/util/feature"
+)
+
+const Name = "colocationprofile"
+
+var (
+	ReconcileInterval    = 30 * time.Second
+	MaxUpdatePodQPS      = 5.0
+	MaxUpdatePodQPSBurst = 10
+)
+
+type Reconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+
+	rateLimiter    *rate.Limiter
+	podUpdateCache gocache.Cache
+}
+
+func newReconciler(mgr ctrl.Manager) *Reconciler {
+	return &Reconciler{
+		Client:         mgr.GetClient(),
+		Recorder:       mgr.GetEventRecorderFor(Name),
+		Scheme:         mgr.GetScheme(),
+		rateLimiter:    rate.NewLimiter(rate.Limit(MaxUpdatePodQPS), MaxUpdatePodQPSBurst),
+		podUpdateCache: *gocache.New(2*ReconcileInterval, 5*time.Minute),
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	profile := &configv1alpha1.ClusterColocationProfile{}
+	if err := r.Client.Get(ctx, req.NamespacedName, profile); err != nil {
+		if !errors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get tenantProfile", "profile", req.NamespacedName)
+			return ctrl.Result{Requeue: true}, err
+		}
+		// not found
+		return ctrl.Result{}, nil
+	}
+
+	if profile.DeletionTimestamp != nil { // skip for a terminating profile
+		return ctrl.Result{}, nil
+	}
+
+	podList, err := r.listPodsForProfile(profile)
+	if err != nil {
+		klog.ErrorS(err, "failed to list pods for ClusterColocationProfile", "profile", profile.Name)
+		return ctrl.Result{Requeue: true}, err
+	}
+	if podList == nil || len(podList.Items) <= 0 {
+		klog.V(6).InfoS("list no pod for ClusterColocationProfile, retry later", "profile", profile.Name)
+		return ctrl.Result{RequeueAfter: ReconcileInterval}, nil
+	}
+	klog.V(5).InfoS("list pods for ClusterColocationProfile", "profile", profile.Name, "pods", len(podList.Items))
+
+	summary := newSummary(profile.Name)
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		// NOTE: Only handle pending and unscheduled pods.
+		if pod.Spec.NodeName != "" || pod.Status.Phase != corev1.PodPending {
+			continue
+		}
+		skip, err := shouldSkipProfile(profile)
+		if err != nil {
+			klog.ErrorS(err, "failed to check skip profile for pod", "profile", profile.Name, "pod", klog.KObj(pod))
+			return ctrl.Result{Requeue: true}, err
+		}
+		if skip {
+			summary.Skipped++
+			klog.V(5).InfoS("skip update Pod by clusterColocationProfile", "profile", profile.Name, "pod", klog.KObj(pod))
+			continue
+		}
+		_, exists := r.podUpdateCache.Get(string(pod.UID))
+		if exists {
+			summary.Cached++
```

Comment:
If the profile is updated frequently, will desired be missed here? Is profile.resourceVersion+Uid used as the update record?

Reply:
good idea!

Reply:
fixed